diff.py

```diff
@@ -414,6 +429,251 @@ def get_arch(arch_str: str) -> "ArchSettings":
+        symbol_formats: Dict[str, Format] = {}
+
+        def symbol_format(s: str) -> Format:
+            # TODO: it would be nice to use a unique Format for each symbol, so we could
+            # add extra UI elements in the HTML version
+            try:
+                f = symbol_formats[s]
+            except KeyError:
+                f = COLOR_ROTATION[
+                    (len(symbol_formats) + base_index) % len(COLOR_ROTATION)
+                ]
+                symbol_formats[s] = f
+            return f
+
+        return symbol_format
+
+    def fields(
+        self,
+        pat: Pattern[str],
+        out1: str,
+        out2: str,
+        color1: FormatFunction,
+        color2: Optional[FormatFunction] = None,
+    ) -> Tuple[str, str]:
+        diffs = [
+            of.group() != nf.group()
+            for (of, nf) in zip(pat.finditer(out1), pat.finditer(out2))
+        ]
+
+        it = iter(diffs)
+
+        def maybe_color(color: FormatFunction, s: str) -> str:
+            return self.apply(s, color(s)) if next(it, False) else s
+
+        out1 = pat.sub(lambda m: maybe_color(color1, m.group()), out1)
+        it = iter(diffs)
+        out2 = pat.sub(lambda m: maybe_color(color2 or color1, m.group()), out2)
+
+        return out1, out2
+
+
+@dataclass
+class PlainFormatter(Formatter):
+    column_width: int
+
+    def apply(self, line: str, f: Format) -> str:
+        return line
+
+    def table(
+        self, header: Optional[Tuple[str, ...]], lines: List[Tuple[str, ...]]
+    ) -> str:
+        if header:
+            lines = [header] + lines
+        return "\n".join(
+            "".join(x.ljust(self.column_width) for x in line) for line in lines
+        )
+
+
+@dataclass
+class AnsiFormatter(Formatter):
+    column_width: int
+
+    def apply(self, line: str, f: Format) -> str:
+        if f == Format.NONE:
+            return line
+        return f"{f.ansi_code}{line}{Style.RESET_ALL}"
+
+    def table(
+        self, header: Optional[Tuple[str, ...]], lines: List[Tuple[str, ...]]
+    ) -> str:
+        if header:
+            lines = [header] + lines
+        return "\n".join("".join(self.ansi_ljust(x) for x in line) for line in lines)
+
+    def ansi_ljust(self, s: str) -> str:
+        """Like s.ljust(width), but accounting for ANSI colors."""
+        needed: int = self.column_width - ansiwrap.ansilen(s)
+        if needed > 0:
+            return s + " " * needed
+        else:
+            return s
+
+
+@dataclass
+class HtmlFormatter(Formatter):
+    def apply(self, line: str, f: Format) -> str:
+        if f == Format.NONE:
+            return line
+        class_name = f.name.lower().replace("_", "-")
+        return f"<span class='{class_name}'>{line}</span>"
+
+    def table(
+        self, header: Optional[Tuple[str, ...]], lines: List[Tuple[str, ...]]
+    ) -> str:
+        if header:
+            lines = [header] + lines
+        # TODO: Make a prettier stylesheet (maybe include it externally)
```

Comment:
We shouldn't force a particular stylesheet on consumers of this, which may clash with their existing designs and be impossible to modify. An external file may be okay as long as we don't emit a `<link rel="stylesheet">` for it. But I wouldn't mind leaving this out of the repo until we make a usable html diff tool that's part of this repo, if we ever do. (I imagine such a tool would support watch mode using websockets or something)

Reply:
Yeah 100% agreed.